test/jdk/java/net/httpclient/HttpRequestNewBuilderTest.java

```diff
@@ -129,190 +139,252 @@ void assertBodyPublisherEqual(HttpRequest r1, HttpRequest r2) {
-       assertTrue(r.bodyPublisher().isPresent());
-       assertEquals(r.bodyPublisher().get().contentLength(), 8);
-       assertAllOtherElementsEqual(r, request, "method");
-
-       testBodyPublisher("PUT", request);
-   }
-
-   @Test(dataProvider = "testRequests")
-   public void testUserDefinedMethod(HttpRequest request) {
-       testBodyPublisher("TEST", request);
-   }
-
-   @Test
-   public void testInvalidMethod() throws URISyntaxException {
-       URI testURI = new URI("http://www.foo.com/");
-       var r = new HttpRequest() {
-           @Override
-           public Optional<BodyPublisher> bodyPublisher() { return Optional.empty(); }
-           @Override
-           public String method() { return "CONNECT"; }
-           @Override
-           public Optional<Duration> timeout() { return Optional.empty(); }
-           @Override
-           public boolean expectContinue() { return false; }
-           @Override
-           public URI uri() { return testURI; }
-           @Override
-           public Optional<Version> version() { return Optional.empty(); }
-           @Override
-           public HttpHeaders headers() { return HttpHeaders.of(Map.of(), (x,y) -> true); }
-       };
-       assertThrows(IAE, () -> HttpRequest.newBuilder(r).build());
-   }
-
-   @Test
-   public void testInvalidURI() throws URISyntaxException {
-       // invalid URI scheme
-       URI badURI = new URI("ftp://foo.com/somefile");
-       var r = new HttpRequest() {
-           @Override
-           public Optional<BodyPublisher> bodyPublisher() { return Optional.empty(); }
-           @Override
-           public String method() { return "GET"; }
-           @Override
-           public Optional<Duration> timeout() { return Optional.empty(); }
-           @Override
-           public boolean expectContinue() { return false; }
-           @Override
-           public URI uri() { return badURI; }
-           @Override
-           public Optional<Version> version() { return Optional.empty(); }
-           @Override
-           public HttpHeaders headers() { return HttpHeaders.of(Map.of(), (x,y) -> true); }
-       };
-       assertThrows(IAE, () -> HttpRequest.newBuilder(r).build());
-   }
+    void assertHeadersEquals(HttpRequest r1, HttpRequest r2, BiPredicate<String, String> filter) {
+        var s1 = r1.headers().map().entrySet().stream();
+        var s2 = r2.headers()
+                .map()
+                .entrySet()
+                .stream()
+                .filter(e -> {
+                    var n = e.getKey();
+                    for (var v : e.getValue()) {
+                        if (filter.test(n, v))
+                            return true;
+                    }
+                    return false;
+                });
+        Iterator<?> iter1 = s1.iterator(), iter2 = s2.iterator();
+        while (iter1.hasNext() && iter2.hasNext())
+            assertEquals(iter1.next(), iter2.next());
+    }
```

Comment:
This code doesn't seem to be correct. It won't work if the `BiPredicate` vetoes only 1 value for a multi-valued request. This lets me think that this scenario is not tested, or if it is tested, then it's not tested with this method.
A simpler implementation would be to rely on map equality. That should work as both map are case insensitive.
That is something like:
```assertEquals(r1.headers().map(), HttpHeaders.of(r2.headers().map(), filter));```
assuming r1 is the request in which headers have been removed according to `filter`.

Reply:
Or if you don't want to trust `HttpHeaders.of` then you should probably use a `record` and `mapMulti` to build a stream of `(name, value)` pairs and then reassemble the map in the end (possibly using `groupingBy`).

Reply:
I removed the stream check and replaced it with an assert based on map equality. I've also added more test cases to check the scenarios suggested above. See https://github.com/openjdk/jdk/pull/1059/commits/03b13e7aa9ff7bc222bd0c466b223ec522ff2973